Issue 25 attribution change

brandonrosenbloom opened the pull request:
SEVERAL CHANGES:
1) Added .idea files to gitignore to help keep the repo clean
2) Added several models to the Resources app in order to help better organize the data while also providing future adjustability.
3) Created a basic author form so new authors could be created while also adding the necessary admin settings. STILL NOT DONE.

aeltanawy committed:
19: Improve installation instructions

-Added extra installation instructions for Mac users
-Added database setup using migrate command

aeltanawy committed:
Revert "19: Improve installation instructions" to add co-authors

This reverts commit 87d35cdf7ccfa1d8245f40694876148a36501e97.

aeltanawy committed:
19: Improve installation instructions

Co-authored-by: Martin Breuss breuss.martin@gmail.com

-Added extra installation instructions for Mac users
-Added database setup using migrate command

meg-ray committed:
adding static resources

meg-ray committed:
Update to getting started page

ewdurbin committed:
force https on heroku

meg-ray committed:
Merge pull request #22 from aeltanawy/master

19: Improve installation instructions

Commit:
Added .idea to the gitignore file

Commit:
Added new Author model and created the necessary forms to manage Author data.

Commit:
Updated DB modeling to have specific tables/models for information like resource status, devices, audiance, etc. Did this so as new values were required, we wouldn't require a code change/redeployment of the application.

Commit:
Admin file/model changes

Registered new models with admin. Created new migration. Added __str__ method to all new models. Updated resource list view to return all resources for now. Will need to update to only allow for approved resources in the future.

Commit:
Update models.py

Commented out language field as it was replaced by a many to many field.

Commit:
Deleted choices file

Deleted the choices file and removed references in models.py, views.py and commented references in tests.py. Need to update the tests to utilize the appropriate models instead of the choices file.

Commit:
Update models.py

Removed the multiselectfield import as it was no longer being used.

Commit:
Resource Status Sequencing

Added a sequence field to the ResourceStatus model so we can programmatically identify the Proposed status without having to rely on the "Proposed" value itself. This will allow us to change the status name if needed.

Commit:
Resource Status Default Logic

Added logic to default the status of a resource to the first status in the sequence so we can change the name of the status in the future if needed.

Commit:
Code Cleanup

Removed unnecessary comments and fields that no longer exist on the Resource model.

Commit:
Model/Admin updates

Updated model name to remove pluralization. Added more details admin capabilities for all models.

Commit:
Form update

Created CreateResourceForm in forms.py and updated views.py to utilize the new custom form. This form will more appropriately handle multiselect fields. Still need to add the ability to add a new author.

Commit:
Update resource_list.html

Updated resource list form to utilize the newly created fields/models. Also appears to have fixed the filters functionality as well.

sijanonly commented on python-in-edu/resources/models.py:
```diff
@@ -4,24 +4,52 @@
-from . import choices
 
 
+def get_initial_status():
```
This is raising issue while migration. I think the issue is because of the fact that `ResourceStatus` won't exist until this point?
I have attached the screenshot herewith. 
![resource](https://user-images.githubusercontent.com/4557587/120923055-bf4c2a80-c6cc-11eb-9ece-f5dc61d76e2b.jpg)

sijanonly commented on python-in-edu/resources/models.py:
```diff
@@ -4,24 +4,52 @@
+        return f'{self.name}'
+
+
+class ProfilePopulation(models.Model):
```
I think it's better to have a `fixture` file so that we can run `loaddata` to load all the possible values to models `ProfileRole`, `ProfilePopulation` and similar models.
I tried to register, and it raises ` NOT NULL constraint failed: resources_profile.populations_id` and I was not aware that `Profile` table has association with `ProfilePopulation` table, and it is empty.

brandonrosenbloom replied:
Instead of using a fixture file, it seems that most folks prefer the use of Data Migrations to initialize data in the DB since they are automatically included in the django migrate operations. I'll start working on that instead.

sijanonly replied:
Okay. That sounds more handy than having a separate dump process. So, that will be called during db migrations itself?

brandonrosenbloom replied:
Correct. According to the django docs, the fixtures are really used from a testing perspective whereas initial data loading should be handled via data migrations.

sijanonly replied:
That makes perfect sense. Thanks for pointing it out.

Commit:
Create 0014_auto_20210607_1959.py

Created a data migration file that will be used to automatically populate several tables within the DB.

brandonrosenbloom commented:
@sijanonly I have now added the data migration file. You should be able to pull the latest version of this request down and run migrate to see the data appear.

sijanonly commented:
> @sijanonly I have now added the data migration file. You should be able to pull the latest version of this request down and run migrate to see the data appear.

@brandonrosenbloom I am still getting this issue while running migration. 

can you please have a look?


![mig_issue](https://user-images.githubusercontent.com/4557587/121243760-63d19680-c89e-11eb-9d6e-5fc20b206800.png)

Commit:
Update models.py

Updated get_initial_status to return None if there is no default status available.

Commit:
Migration update

Removed unneeded period.

brandonrosenbloom commented:
@sijanonly Try now. I've updated the get_default_value method to return a none if there are no values available.

sijanonly commented:
> @sijanonly Try now. I've updated the get_default_value method to return a none if there are no values available.

@brandonrosenbloom , the changes didn't work. Actually, the statement `ResourceStatus.objects.get(sequence=1).id` will raise error since it won't find the ResourceStatus object. I changed it to following and it works.

```
def get_initial_status():
    try:
        initial_status = ResourceStatus.objects.get(sequence=1).id
    except ResourceStatus.DoesNotExist:
        initial_status = None
    return initial_status
```

Also, there is one issue I am facing while creating user. I tried to create superuser     `python manage.py createsuperuser`
and `Profile` model has `populations` field, which is NOT NULL. And it raises 
`django.db.utils.IntegrityError: NOT NULL constraint failed: resources_profile.populations_id`

I think it's better to set `null=True` for  `populations` field in `Profile` model.

Commit:
Update models.py

Added try/except logic to get_initial_status method as recommended by @sijanonly

brandonrosenbloom commented:
@sijanonly I've applied the change to the get_initial_status method as suggested. @meg-ray Do we want to guarantee that all profiles have a population entered?

sijanonly commented:
> @sijanonly I've applied the change to the get_initial_status method as suggested. @meg-ray Do we want to guarantee that all profiles have a population entered?

> @brandonrosenbloom I just found that there is a separate `ProfileUpdateView` where user can update `populations`,.. information. It seems like we can put `None` for default ?

brandonrosenbloom commented:
@sijanonly The only issue I see with setting the population attribute to None on the model is really how it may impact the business process by which users are created. @meg-ray if we don't expect new users to always be created with a population value, then setting the value to no longer be required makes sense. However, if we expect all new users to have a population to be identified when they are first created, then we should probably leave the field as required. Thoughts?
